packages/contracts/contracts/optimistic-ethereum/ovm/StateTransitioner.sol

```diff
@@ -1,142 +1,324 @@
-            "0x12",
-            0x1212121212121212121212121212121212121212,
-            0x1212121212121212121212121212121212121212,
-            false
+            transactionData.timestamp,
+            transactionData.queueOrigin,
+            transactionData.ovmEntrypoint,
+            transactionData.callBytes,
+            transactionData.fromAddress,
+            transactionData.l1MsgSenderAddress,
+            transactionData.allowRevert
+        );
+
+        require(
+            stateManager.existsInvalidStateAccessFlag() == false,
+            "Detected an invalid state access."
         );
-        require(stateManager.existsInvalidStateAccessFlag() == false, "Detected invalid state access!");
-        currentTransitionPhase = TransitionPhases.PostExecution;
 
-        // This will allow people to start updating the state root!
-        return true;
+        currentTransitionPhase = TransitionPhases.PostExecution;
     }
 
-    /****************************
-    * Post-Transaction Execution *
-    ****************************/
-    function proveUpdatedStorageSlot() public postExecutionPhase {
+    /******************************
+     * Post-Transaction Execution *
+     ******************************/
+
+    /**
+     * @notice Updates the root of the state trie by making a modification to
+     * a contract's storage slot. Contract storage to be modified depends on a
+     * stack of slots modified during execution.
+     * @param _stateTrieWitness Merkle trie inclusion proof for the contract
+     * within the current state trie.
+     * @param _storageTrieWitness Merkle trie inclusion proof for the slot
+     * being modified within the contract's storage trie.
+     */
+    function proveUpdatedStorageSlot(
+        bytes memory _stateTrieWitness,
+        bytes memory _storageTrieWitness
+    ) public onlyDuringPhase(TransitionPhases.PostExecution) {
         (
             address storageSlotContract,
             bytes32 storageSlotKey,
             bytes32 storageSlotValue
         ) = stateManager.popUpdatedStorageSlot();
-        // TODO: Prove inclusion / make this update to the root
+
+        stateRoot = ethMerkleTrie.updateAccountStorageSlotValue(
+            storageSlotContract,
+            storageSlotKey,
+            storageSlotValue,
+            _stateTrieWitness,
+            _storageTrieWitness,
+            stateRoot
+        );
     }
-    function proveUpdatedContract() public postExecutionPhase {
+
+    /**
+     * @notice Updates the root of the state trie by making a modification to
+     * a contract's state. Contract to be modified depends on a stack of
+     * contract states modified during execution.
+     * @param _stateTrieWitness Merkle trie inclusion proof for the contract
+     * within the current state trie.
+     */
+    function proveUpdatedContract(
+        bytes memory _stateTrieWitness
+    ) public onlyDuringPhase(TransitionPhases.PostExecution) {
```

Comment:
I don't believe an ordering requirement is necessary. The storage trie root is updated separately from the rest of the account state (we can selectively update specific fields because the account state is RLP encoded). This function is effectively limited to updating the nonce of a contract or, in the case of a contract creation, inserting the initial state. I might be missing a case in which ordering is necessary but I can't think of one right now.